core/src/cluster_slot_state_verifier.rs

```diff
@@ -30,151 +223,218 @@ pub enum ResultingStateChange {
-            // If the versions do not match, modify fork choice rule
-            // to exclude our version from being voted on and also
-            // repair correct version
-            warn!(
-                "Cluster duplicate_confirmed slot {} with hash {}, but we froze slot with hash {}",
-                slot, cluster_duplicate_confirmed_hash, bank_frozen_hash
-            );
-            return vec![
-                ResultingStateChange::MarkSlotDuplicate(*bank_frozen_hash),
-                ResultingStateChange::RepairDuplicateConfirmedVersion(
-                    *cluster_duplicate_confirmed_hash,
-                ),
-            ];
-        } else {
-            // If the versions match, then add the slot to the candidate
-            // set to account for the case where it was removed earlier
-            // by the `on_duplicate_slot()` handler
-            return vec![ResultingStateChange::DuplicateConfirmedSlotMatchesCluster(
-                *bank_frozen_hash,
-            )];
+    let bank_hash = bank_status.bank_hash().expect("bank hash must exist");
+    let is_dead = bank_status.is_dead();
+
+    let mut state_changes = vec![];
+    check_duplicate_confirmed_hash_against_frozen_hash(
+        &mut state_changes,
+        slot,
+        duplicate_confirmed_hash,
+        bank_hash,
+        is_dead,
+    );
+
+    state_changes
+}
+
+fn on_duplicate(duplicate_state: DuplicateState) -> Vec<ResultingStateChange> {
+    let DuplicateState {
+        bank_status,
+        duplicate_confirmed_hash,
+    } = duplicate_state;
+
+    match bank_status {
+        BankStatus::Dead | BankStatus::Frozen(_) => (),
+        // No action to be taken yet
+        BankStatus::Unprocessed => {
+            return vec![];
         }
     }
 
-    if is_slot_duplicate {
-        // If we detected a duplicate, but have not yet seen any version
-        // of the slot duplicate_confirmed (i.e. block above did not execute), then
-        // remove the slot from fork choice until we get confirmation.
+    let bank_hash = bank_status.bank_hash().expect("bank hash must exist");
 
-        // If we get here, we either detected duplicate from
-        // 1) WindowService
-        // 2) A gossip duplicate_confirmed version that didn't match our frozen
-        // version.
-        // In both cases, mark the progress map for this slot as duplicate
-        return vec![ResultingStateChange::MarkSlotDuplicate(*bank_frozen_hash)];
+    // If the cluster duplicate_confirmed some version of this slot that matches our
+    // version, ignore this duplicate signal
+    if let Some(duplicate_confirmed_hash) = duplicate_confirmed_hash {
+        if duplicate_confirmed_hash == bank_hash {
+            return vec![];
+        }
     }
 
-    vec![]
+    // If we either:
+    // 1) Have not yet seen any version of the slot duplicate_confirmed,
+    // 2) Our version of the slot does not match the duplicate_confirmed version,
```

Comment:
Yeah I was thinking about this too, because if the slot is duplicate confirmed and our frozen hash doesn't match, the duplicate flag should already have been set by either the `Frozen` or the `DuplicateConfirmed`  state transitions.

There is no correctness issue for called the mark duplicate multiple times on the same node, it just incurs this cost: https://github.com/solana-labs/solana/blob/40f207546f5fbdee0c5a7293f0b788a810da1e14/core/src/heaviest_subtree_fork_choice.rs#L983-L1008. I guess this can be avoided in this situation.

Updated!